File: c/goertzel.c
/* -*- c-file-style:"stroustrup"; indent-tabs-mode: nil -*- */

#include "goertzel.h"

#include <math.h>


struct GoertzelFilter makeGoertzelFilter(double f, double fs)
{
    struct GoertzelFilter rslt;
    rslt.f = f;
    rslt.fs = fs;
    rslt.koef = GoertzelFilter_koef(f, fs);
    GoertzelFilter_Reset(&rslt);

    return rslt;
}

    
void GoertzelFilter_Reset(struct GoertzelFilter *pf)
{
    pf->vn._1 = pf->vn._2 = 0.0;
}


struct GoertzelFilter_Vn GoertzelFilter_kernel(double sample[], size_t n, double k, struct GoertzelFilter_Vn vn)
{
    size_t i;
    for (i = 0; i < n; ++i) {
        double t = k * vn._1 - vn._2 + sample[i];
        vn._2 = vn._1;
        vn._1 = t;
    }
    return vn;
}
    

double GoertzelFilter_power(double koef, struct GoertzelFilter_Vn vn, int n)
{
    double rslt = vn._1*vn._1 + vn._2*vn._2 - koef * vn._1 * vn._2;
    if (rslt < EPSILON) {
        rslt = EPSILON;
    }
    return rslt / (n*n);
}
 

double calc_dBm(double power)
{
    return 10 * log10(2 * power * 1000 / 600.0);
}
    

double GoertzelFilter_koef(double f, double fs)
{
    return 2 * cos(2 * M_PI * f / fs);
}
    

double GoertzelFilter_process(struct GoertzelFilter *pGF, double samples[], size_t n)
{
    pGF->vn = GoertzelFilter_kernel(samples, n, pGF->koef, pGF->vn);
    return GoertzelFilter_power(pGF->koef, pGF->vn, n);
}

File: rust/goertzel.rs
fn calc_koef(f: f64, fs: f64) -> f64 {
    assert!(f < fs / 2.0);
    assert!(fs > 0.0);
    2.0 * (2.0 * std::f64::consts::PI * f / fs).cos()
}

/// Helper for remembering the last two filter "running values" for
/// the Geortzel filter implemented as an IIR filter.
#[derive(Copy, Clone, Debug)]
pub struct Vn {
    _1: f64,
    _2: f64
}

/// Holds data for a Goertzel filter
#[derive(Debug)]
pub struct Filter {
    /// The frequency of the filter
    f: f64,
    /// The sampling frequency
    fs: f64,
    /// The Goertzel coefficient, calulcated
    /// from the above mentioned frequencies
    koef: f64,
    /// The running values of the Goertzel filter calculation.
    vn: Vn
}

impl Filter {
    /// Returns a Goertzel filter for the given parameters
    /// * `f` The frequency of the filter
    /// * `fs`  The sampling frequency of the samples to process
    pub fn new(f: f64, fs: f64) -> Filter {
	Filter {
	    f: f,
	    fs: fs,
	    koef: calc_koef(f, fs),
	    vn: Vn{_1: 0.0, _2: 0.0}
	}
    }
    
    /// Resets the filter so that we can start it over again.
    pub fn reset(&mut self) {
	self.vn._1 = 0.0;
	self.vn._2 = 0.0;
    }
    
    /// Process the samples using the filter.
    /// Returns the resulting power of the signal at the filter frequency
    pub fn process(&mut self, sample: &[f64]) -> f64 {
	kernel(sample, self.koef, &mut self.vn);
	power(self.koef, self.vn, sample.len())
    }
}

/// The "kernel" of the Gortzel filter as an IIR filter
pub fn kernel(sample: &[f64], k: f64, vn: &mut Vn) {
    for x in sample.iter() {
	let t = k * vn._1 - vn._2 + x;
	vn._2 = vn._1;
	vn._1 = t;
    }
}

/// Returns the power of the signal that has passed through a Goertzel
/// filter.
pub fn power(k: f64, vn: Vn, n: usize) -> f64 {
    let mut rslt = vn._1 * vn._1 + vn._2 * vn._2 - k * vn._1 * vn._2;
    if rslt < f64::EPSILON  {
	rslt = f64::EPSILON;
    }
    rslt / (n*n) as f64
}

/// Returns the dBm of the given power of a signal
pub fn dbm(power: f64) -> f64 {
    10.0 * (2.0 * power * 1000.0 / 600.0).log10()
}
